src/prefect/core/flow.py

```diff
@@ -883,21 +883,26 @@ def _run_on_schedule(
 
             ## create next scheduled run
             try:
-                cached_tasks = {
-                    t: s for t, s in flow_state.result.items() if s.is_cached()
-                }
+                cached_tasks = {}
+
                 for t, s in flow_state.result.items():
                     if s.is_cached():
                         cached_tasks.update({t: s})
                     elif s.is_mapped() and any(
-                        [sub_state.is_cached() for sub_state in s.map_states]
+                        sub_state.is_cached() for sub_state in s.map_states
                     ):
-                        new = prefect.engine.state.Mapped(map_states=[])
-                        for sub_state in s.map_states:
-                            new.map_states.append(
-                                sub_state if sub_state.is_cached() else None
-                            )
-                        cached_tasks.update({t: new})
+                        cached_sub_states = [
+                            sub_state
+                            for sub_state in s.map_states
+                            if sub_state.is_cached()
```

Comment:
this should have an `else None` condition for those mapped states which are _not_ cached, to ensure that the order / length of the mapped states is consistent.

Reply:
Nice catch. Fixed. One question...why is this not caught by the test? Would it make sense to add something around this?

Reply:
Yea, this change only affects the situation where some of the "mapped children" are cached and some aren't - my test has _both_ cached.  I will add a test for this!